### PR TITLE
feat: support Qwen3.5 with verl 0.8.0 and flash-attn CUDA 13

### PR DESCRIPTION
**Pull request by nashjojo**
## Summary

Adds end-to-end support for Qwen3.5-4B training by upgrading to the latest trinity-rft, verl, transformers, and flash-attn stack.

### Dependency upgrades (`pyproject.toml`)
- `trinity-rft[vllm]>=0.5.2`
- `verl>=0.8.0`
- `transformers>=5.2,<6`
- `huggingface-hub>=1.3`
- Added `numpy` to dependencies
- Added `[tool.uv]` override-dependencies for verl to resolve transitive pin conflicts

### FSDP training backend (`fsdp_training_backend.py`)
- Fix temperature tensor shape: `torch.full((B,1,1), 1.0)` → `torch.ones(B)` for verl 0.8.0 compatibility
- Call `torch.cuda.set_device()` before `dist.init_process_group` to avoid multi-GPU NCCL issues
- Add `async_init` logging for better observability during FSDP cold start

### Sampling backend (`sampling_backend.py`)
- Set `VLLM_RAY_PER_WORKER_GPUS=0` in Ray runtime env to prevent vLLM v1 engine from over-allocating GPUs

### flash-attn installer (`scripts/install_flash_attn.py`)
- Update to flash-attn 2.8.3
- Add CUDA 13 wheel detection and installation support

## Files Changed
- `pyproject.toml` (+16/-6)
- `scripts/install_flash_attn.py` (+15/-10)
- `src/tuft/backends/fsdp_training_backend.py` (+35/-19)
- `src/tuft/backends/sampling_backend.py` (+4/-1)

## Test Plan

- [x] Qwen3.5-4B 32k SFT training on single A100-80GB (6/6 runs passed)
- [x] Pre-commit checks pass (ruff, pyright, detect-secrets, pytest)
- [x] flash-attn installs correctly on CUDA 13 environment

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -11,14 +11,14 @@ requires-python = ">=3.11"
 dependencies = [
     "fastapi>=0.125.0",
     "httpx>=0.28.1",
-    "numpy<2.0.0",
```
Comment (ekzhu) [inline pyproject.toml:14]: Do we still need a lower/upper bound?

File: pyproject.toml [hunk 2]
```diff
@@ -125,6 +127,18 @@ ruff = ["--fix"]
...
+# Override upstream constraints that lag behind our requirements:
+# - Qwen3.5 needs transformers>=5.2 for architecture registration.
+# - vLLM 0.23+ for Qwen3.5 support; verl 0.8.0 supports vLLM>=0.8.2.
+# - numpy>=2 needed by vllm/opencv-python-headless.
+[tool.uv]
+override-dependencies = [
+    "transformers>=5.2.0,<6.0.0",
+    "vllm>=0.16.0",
+    "numpy>=2.0.0",
+    "verl>=0.8.0",
```
Comment (ekzhu) [inline pyproject.toml]: No upper bound on `vllm`/`verl`, but `worker_patch.py` rejects vLLM >0.23.0 — a fresh resolve to 0.24+ hard-fails inside Ray actors. Pin `vllm>=0.20,<0.24` and `verl>=0.8,<0.9`. The v23 logprobs path only activates at >=0.20, so the `>=0.16.0` floor is also too low.

File: pyproject.toml [hunk 3]
```diff
@@ -11,14 +11,14 @@ requires-python = ">=3.11"
 dependencies = [
     "fastapi>=0.125.0",
     "httpx>=0.28.1",
-    "numpy<2.0.0",
+    "numpy",
```
Comment (ekzhu) [inline pyproject.toml]: `<2.0.0` cap dropped and forced to `>=2` via the override. Worth confirming tinker tolerates numpy 2 — install CI won't catch a training-time incompat.

File: src/tuft/patches/apply.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                # Compare contents
+                if src_file.read_text() != dest.read_text():
+                    patched.append(f"[NEEDS UPDATE] {dest}")
+                else:
+                    patched.append(f"[OK] {dest}")
+            else:
+                patched.append(f"[MISSING TARGET] {dest}")
+        else:
+            shutil.copy2(src_file, dest)
+            patched.append(str(dest))
```
Comment (ekzhu) [inline src/tuft/patches/apply.py]: Copies regardless of whether `dest` exists, so if trinity renames/moves these modules the patch silently creates orphan files and no-ops at runtime. Mirror the `--check` `[MISSING TARGET]` guard and fail/warn when `dest` is absent.

File: scripts/install.sh [hunk 5]
```diff
@@ -202,6 +202,10 @@ install_tuft() {
     # Install tuft with all extras (backend, persistence)
     uv pip install --python "$TUFT_VENV/bin/python" "${PACKAGE_SPEC}[backend,persistence]"
 
+    # Apply trinity-rft compatibility patches for vLLM >= 0.20
+    print_step "Applying trinity-rft vLLM patches..."
+    "$TUFT_VENV/bin/python" -m tuft.patches.apply || print_warning "Failed to apply trinity patches (non-fatal)"
```
Comment (ekzhu) [inline scripts/install.sh]: `|| print_warning` makes a failed patch non-fatal; combined with apply.py copying regardless of target, a silently-broken patch only surfaces at training time. Prefer failing here.

File: src/tuft/backends/fsdp_training_backend.py [hunk 6]
```diff
@@ -218,12 +218,9 @@ def _datum_list_to_tensordict(
...
         "loss_mask": weights_device,
-        # Temperature shape (B, 1, 1): TensorDict requires all leaf tensors to
-        # have batch_size as a prefix dimension. A 0-dim scalar is rejected by
-        # td batch_size validation, and verl's use_remove_padding=True path does
-        # `logits.div_(temperature)` which broadcasts (B,1,1) -> (B, seq, V).
-        # Keep (B,1,1) — it is correct for both padded and rmpad paths.
-        "temperature": torch.full((batch_size, 1, 1), 1.0, device=device, dtype=torch.float32),
+        # Temperature: verl 0.8.0 expects (bsz,) and does unsqueeze internally.
+        # (verl 0.7.x expected (B,1,1) for direct broadcast, but 0.8.0 changed.)
+        "temperature": torch.ones(batch_size, device=device, dtype=torch.float32),
```
Comment (ekzhu) [inline src/tuft/backends/fsdp_training_backend.py]: `(B,)` is correct only because verl 0.8.0 unsqueezes internally — pin `verl<0.9`, else `logits.div_(temperature)` broadcasts wrong silently. Also the `use_remove_padding=False` comment below (~L261) still says temperature "must be (B,1,1)" — now stale, please update.

File: src/tuft/patches/__main__.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+"""Allow running as: python -m tuft.patches.apply"""
```
Comment (ekzhu) [inline src/tuft/patches/__main__.py]: Docstring is off: `__main__.py` enables `python -m tuft.patches`. `python -m tuft.patches.apply` already works via apply.py's own guard.

**Review by ekzhu — COMMENTED** [review]
Compatibility work looks solid and CI is green. A few inline notes — mainly version-pin footguns that fail silently at training time rather than install time.

File: pyproject.toml (hunk 1 above)
Comment (nashjojo) [reply pyproject.toml:14]: 是 commit b7dd407（我们这次的 PR commit）把 "numpy<2.0.0" 改成了 "numpy"。
原因： verl 0.8.0 和 vLLM 0.23.0 都需要 numpy>=2.0.0（通过 opencv-python-headless 等间接依赖）。旧的 numpy<2.0.0 会直接与它们冲突。所以我们：
主依赖里去掉上限："numpy"（让 pip/uv 自由选择兼容版本）
override-dependencies 里加 "numpy>=2.0.0" 强制覆盖其他包（如旧 verl 0.7.1）可能对 numpy 的 <2 限制

现在还需要 override 里的 "numpy>=2.0.0" 吗？
需要保留。虽然我们的直接依赖不再限制 numpy 上限，但 trinity-rft 0.5.2（通过旧 verl 0.7.1）可能间接 pin numpy<2。override 确保即使这些子依赖有 <2 约束，uv 也会安装 numpy 2.x。
如果不加 override，uv 可能为了满足 trinity-rft 的传递约束降级到 numpy 1.x，然后 vLLM 0.23.0 就装不上了。
总结： 当前配置是正确的——主依赖 "numpy"（无限制） + override "numpy>=2.0.0"（强制覆盖子依赖约束）。

**Comment by ekzhu** [comment]
Thanks for the refresh. The old custom-patch comments are moot now, but this is not merge-ready yet:

- A built wheel fails Linux dependency resolution outside the checkout: `[tool.uv]` overrides are root-only/not published (`verl 0.8` requires NumPy <2; vLLM/OpenCV requires >=2). CI masks this by installing from `$GITHUB_WORKSPACE`.
- Multi-GPU FSDP selects CUDA by global rank; each one-GPU Ray actor sees logical `cuda:0`, so rank 1+ will fail before NCCL init.
- The sampling env override is absent without `worker_venv_path`, unused by Trinity 0.6, and overwritten by allowed Trinity 0.5.2. Please pin a coherent stack (at least Trinity >=0.6, Verl <0.9, and a compatible vLLM cap) and update the stale `(B,1,1)` comment.
- Please verify or mirror+checksum the community CUDA 13 wheel; only the cp312/Torch 2.11/x86_64/ABI-true tuple exists.
- `test-upgrade-from-source` now tests `--local-source`, leaving the real path untested.

All current checks are green, but GPU paths are not exercised.

**Comment by nashjojo** [comment]
> Thanks for the refresh. The old custom-patch comments are moot now, but this is not merge-ready yet:
> 
> * A built wheel fails Linux dependency resolution outside the checkout: `[tool.uv]` overrides are root-only/not published (`verl 0.8` requires NumPy <2; vLLM/OpenCV requires >=2). CI masks this by installing from `$GITHUB_WORKSPACE`.
> * Multi-GPU FSDP selects CUDA by global rank; each one-GPU Ray actor sees logical `cuda:0`, so rank 1+ will fail before NCCL init.
> * The sampling env override is absent without `worker_venv_path`, unused by Trinity 0.6, and overwritten by allowed Trinity 0.5.2. Please pin a coherent stack (at least Trinity >=0.6, Verl <0.9, and a compatible vLLM cap) and update the stale `(B,1,1)` comment.
> * Please verify or mirror+checksum the community CUDA 13 wheel; only the cp312/Torch 2.11/x86_64/ABI-true tuple exists.
> * `test-upgrade-from-source` now tests `--local-source`, leaving the real path untested.
> 
> All current checks are green, but GPU paths are not exercised.

    fix(review): make Qwen3.5 stack merge-ready (points 1-5)
    
    1. Published metadata self-consistent: pin verl to the #6551 commit (verl@git, numpy>=2) and DROP the root-only [tool.uv] override; add [tool.hatch.metadata].allow-direct-references. wheel/git installs now resolve (verl 0.9.0.dev0, numpy 2.3.5) without the checkout-only override.
    
    2. Multi-GPU FSDP: each Ray num_gpus=1 actor sees its GPU as cuda:0; use torch.cuda.set_device(0) instead of global self.rank (rank>=1 previously selected a nonexistent device and failed before NCCL init).
    
    3. Coherent stack: trinity-rft[vllm]>=0.6.0, explicit vllm>=0.19.1,<=0.23.0; revert dead VLLM_RAY_PER_WORKER_GPUS env in sampling_backend (Trinity 0.6 sets it itself).
    
    4. flash-attn cu13 wheel: guard the only-existing (cp312/torch2.11/linux_x86_64/abiTRUE) tuple and verify a pinned SHA-256 before install.
    
    5. CI: test-upgrade-from-source now exercises the real VCS path via TUFT_GIT_URL=file://$GITHUB_WORKSPACE (install.sh --from-source URL is overridable).
    
    Verified on Qwen3.5-4B: fsdp=1 full smoke test 6/6; fsdp=2 grad-accum passes across both ranks (Point 2).

**Comment by ekzhu** [comment]
This is much better. Two remaining changes:

- `FLASH_VERSION` was moved globally to 2.8.3, so CUDA 12 installs now request mirror paths that return 403; the equivalent 2.8.1 tuple works. Please keep separate CUDA 12/13 version maps or publish and verify the CUDA 12 artifacts. The failure is optional, so affected installs continue without flash-attn.
- For the temporary Verl workaround, keep normal `verl>=0.8,<0.9` project metadata, restore the root uv override, and have every initial/upgrade path in `install.sh` pass an override file containing the pinned Verl Git requirement. A NumPy-only override misses the NumPy 2 runtime fixes in verl-project/verl#6551. Also ensure Docker's final `uv pip install .[dev]` sees the same override. A normal direct requirement cannot satisfy `<0.9` because the pinned commit reports `0.9.0.dev0`; `--override` can.

**Comment by nashjojo** [comment]
> fix(review): per-CUDA flash-attn version + verl uv-override redesign

fix(review): per-CUDA flash-attn version + verl uv-override redesign
flash-attn: a global bump to 2.8.3 made CUDA 12 request an aliyun path that 403s (only 2.8.1 is published there). Use a per-CUDA map (FLASH_VERSIONS={'12':'2.8.1','13':'2.8.3'}): cu12->2.8.1 aliyun (as before), cu13->2.8.3 GitHub + pinned SHA-256.

verl: keep clean, publishable metadata verl>=0.8.0,<0.9 (drop the direct git URL and [tool.hatch.metadata].allow-direct-references) and force the #6551 commit via uv override instead: restore pyproject [tool.uv] override-dependencies for project-mode installs, and add scripts/verl-git-override.txt passed via 'uv pip install --override' on every install.sh path and Docker's final '.[dev]'. A normal direct requirement can't satisfy <0.9 (the commit reports 0.9.0.dev0) and misses #6551's numpy-2/vLLM/nccl runtime fixes; --override replaces the constraint and carries the fixed commit.

Verified (uv --dry-run): project-mode -e . (via [tool.uv] override) and registry-mode tuft[...] --override FILE both resolve 280 packages with no URL-dependency error and no <0.9 conflict; verl -> 0.9.0.dev0, numpy 2.3.5. flash-attn cu13 2.8.3 re-installs + SHA-256 verified + imports. Runtime stack unchanged from the prior 6/6 GPU validation.
